### ISSUE TITLE
[OSX - 88832e3679b6d630] compile break

### ISSUE THREAD

**Issue by Havner** [dxx-rebirth/dxx-rebirth#353]
```
similar/main/game.cpp:1636:39: error: non-type template argument evaluates to 2147483648, which cannot be narrowed to type 'int'
      [-Wc++11-narrowing]
constexpr std::integral_constant<fix, 0x80000000> flicker_timer_disabled{};

```

0x80000000 is one higher than a maximum value of int32_t (which fix seems to be).

`typedef int32_t fix;		//16 bits int, 16 bits frac`

```
$ g++ --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin

```

**Comment by Havner** [comment]
The commit-id mentioned in the title is the culprit, not the commit I tested this on.

**Comment by Havner** [comment]
![screen shot 2017-11-06 at 00 37 26](https://user-images.githubusercontent.com/6246347/32420447-b7f3f06e-c28a-11e7-805d-d31a05e5bc7f.png)

Not necessarily white, sometime brown. But always garbage.

**Comment by vLKp** [comment]
Interesting.  GNU GCC accepts this without complaint, both for gcc-5 and gcc-7.

**Comment by Havner** [comment]
Still, is this intentional to assign a value outside of its range?

**Comment by vLKp** [comment]
It's always been that way.  The only change in the identified commit is:

```diff
--- a/similar/main/game.cpp
+++ b/similar/main/game.cpp
@@ -1633,7 +1633,7 @@ static void slide_textures(void)
 Flickering_light_array_t Flickering_lights;
 unsigned Num_flickering_lights;
 
-constexpr fix flicker_timer_disabled = 0x80000000;
+constexpr std::integral_constant<fix, 0x80000000> flicker_timer_disabled{};
 
 static void flicker_lights(fvmsegptridx &vmsegptridx)
 {
```

I think it is intentional that it be out of range, since it is used as a magic flag value.  Unfortunately, it cannot be changed to some other magic value now, since the structure member with which it is used is written to the save file.

**Comment by Havner** [comment]
Value inside a variable cannot be out of range. Assigning MAX32_INT + 1 to in32_t means it's value will be basically INT32_MIN. If that is the magic value, fine. But the fix for this is pretty simple.

**Comment by Havner** [comment]
It compiles fine without warnings (in this line at least). I could not check the GDB, as it seems you need to sign GDB binary on OSX to be able to use.

I've wrote a simple test program though and checked it with clang to make sure its behavior is what you expect.

```
#include <type_traits>
#include <iostream>
#include <limits.h>

typedef int32_t fix;		//16 bits int, 16 bits frac

constexpr fix flicker_timer_disabled_old = 0x80000000;
constexpr std::integral_constant<fix, static_cast<fix>(0x80000000)> flicker_timer_disabled_new{};

int main()
{
	std::cout << INT32_MIN << std::endl;
	std::cout << flicker_timer_disabled_old << std::endl;
	std::cout << flicker_timer_disabled_new << std::endl;
	return 0;
}
```

No compilation warning with `g++ -std=c++11 x.cpp -Wall` and the result:
```
-2147483648
-2147483648
-2147483648

```

**Comment by vLKp** [comment]
> Value inside a variable cannot be out of range. Assigning MAX32_INT + 1 to in32_t means it's value will be basically INT32_MIN. If that is the magic value, fine. But the fix for this is pretty simple.

Yes, and the critical point is that the value must be bitwise identical to what it always was, else old savegames break.  I thought `INT32_MIN` would evaluate to the same number, but was not in a position to verify that when I wrote my prior comment.  I was more confident that explicitly narrowing the value would produce the right result, so I proposed that.  Local testing suggests that `INT32_MIN` can be used as a template parameter to get the same result as passing `0x80000000`.

[Fix clang build failure on flicker_timer_disabled](/dxx-rebirth/dxx-rebirth/commit/425b78b8a63b27a37e38cc4911f3a27dc59c0661) should resolve this.

> I could not check the GDB, as it seems you need to sign GDB binary on OSX to be able to use.

What's the point of installing a program you cannot use?

**Comment by Havner** [comment]
You mean installing GDB by apple? It wasn't installed, at least not in $PATH, maybe somewhere inside XCode, although I think they use some sort of LLVM derived debugger. GDB I got from MacPorts and it's them that informed about this limitation.

**Comment by vLKp** [comment]
I see.  In this context, gdb is merely used as a way to print data from the program's data; running the program is not required, so no signature should be necessary.

Does the clang build work for you now?
